### ISSUE TITLE
Error: EPERM: operation not permitted

### ISSUE THREAD

**Issue by piuccio** [isaacs/rimraf#155]
When I run chrome-headless I get the error

```
assert.js:81
  throw new assert.AssertionError({
  ^
AssertionError: false == true
    at rmdir (/Users/fabio/repo/prerender-chrome-headless/node_modules/chrome-launcher/node_modules/rimraf/rimraf.js:159:5)
    at /Users/fabio/repo/prerender-chrome-headless/node_modules/chrome-launcher/node_modules/rimraf/rimraf.js:95:13
    at FSReqWrap.oncomplete (fs.js:114:15)
```

The offending line is [this one](https://github.com/isaacs/rimraf/blob/master/rimraf.js#L217) and the assertion fails because `originalEr` appears not to be an instance of `Error`.

```js
originalEr instanceof Error // false
typeof originalEr // object
originalEr.constructor // function Error() { [native code] }
JSON.stringify(originalEr)  /* {
 "errno":-1,
 "code":"EPERM",
 "syscall":"unlink",
 "path":"/var/folders/g8/blah_blah/T/lighthouse.XXXXXXX.abc"
} */
originalEr.message // EPERM: operation not permitted, unlink '/var/folders/g8blah_blah/T/lighthouse.XXXXXXX.abc'
```

I find surprising that `originalEr instanceof Error` returns `false`, I don't know how to explain that.

You can reproduce it by cloning [this repo](https://github.com/en-japan-air/prerender-chrome-headless) and running `npm test`.

I'm on Mac OSX 10.12.4
Node 7.10.0

It happens on node `6.11.0` and `8.1.2` as well.

I tried to reproduce the error in another repo that doesn't use chrome-headless but I couldn't. The folder it's trying to delete exists, is not empty, has mode `0700` and contains few files and folders, some of them have the `com.apple.rootless` and `com.apple.quarantine` flag applied.

I've also tried running rimraf directly on that folder that causes an error but it gets deleted correctly, so there must be something else in the chrome-headless setup that messes up the permission.

Another annoying thing is that assertion errors are not caught, so the callback is never called with the error and your program is stuck forever.

**Comment by tmeasday** [comment]
I'm seeing this too, using [chromy](https://www.npmjs.com/package/chromy) under some circumstances that I've not been able to isolate.

> I tried to reproduce the error in another repo that doesn't use chrome-headless but I couldn't. The folder it's trying to delete exists, is not empty, has mode 0700 and contains few files and folders, some of them have the com.apple.rootless and com.apple.quarantine flag applied.

I don't think the issue has to do with the folder being deleted. `rimraf` works by first trying to unlink the directory as a file, then catches the subsequent error (this one) and instead does a proper recursive `rm -r` inside. So the error isn't the problem; the problem is that it's a `Error` than isn't an `instanceof Error`.

**Comment by tmeasday** [comment]
@piuccio -- I'm pretty sure the problem is somehow due to something funky lighthouse is doing and given that it uses rimraf as a library, it's fair to say it's not really rimraf's fault. I've taken the liberty of quoting you at https://github.com/GoogleChrome/lighthouse/issues/2618 -- hopefully that will help track down the problem.

**Comment by paulirish** [comment]
Thanks for filing these ya'll. Looks like whatever the issue was, it was fixed between rimraf@2.2.8 and rimraf@2.3.0. 

This can be closed. :)
